File: 2016/src/day04.ts
function part1(data: string): number {
  let sum = 0
  for (const l of data.split('\n')) {
    const s = l.split('-')
    const name = s.slice(0, -1).join('')
    const [id, check] = s[s.length - 1].slice(0, -1).split('[')
    const freq: Record<string, number> = {}
    for (const c of name) freq[c] = (freq[c] || 0) + 1
    const checked = Object.entries(freq)
      .sort(([k1, v1], [k2, v2]) => v2 - v1 || k1.localeCompare(k2))
      .map(([k]) => k)
      .slice(0, 5)
      .join('')
    if (checked == check) sum += parseInt(id)
  }
  return sum
}

function part2(data: string): number {
  for (const l of data.split('\n')) {
    const s = l.split('-')
    const id = parseInt(s[s.length - 1].split('[')[0])
    const a = 'a'.charCodeAt(0)
    const name = s
      .slice(0, -1)
      .map((s) =>
        s
          .split('')
          .map((c) =>
            String.fromCharCode(((c.charCodeAt(0) - a + id) % 26) + a)
          )
          .join('')
      )
      .join(' ')
    if (name == 'northpole object storage') return id
  }
  return 0
}

export { part1, part2 }

File: 2024/src/aoc2024/day04.py
def part1(data: str) -> int:
    rows = data.splitlines()
    return sum(
        1
        for i in range(len(rows))
        for j in range(len(rows[0]))
        if rows[i][j] == "X"
        for u in (-1, 0, 1)
        if 0 <= i + 3 * u < len(rows)
        for v in (-1, 0, 1)
        if 0 <= j + 3 * v < len(rows[0])
        if rows[i + u][j + v] == "M"
        if rows[i + 2 * u][j + 2 * v] == "A"
        if rows[i + 3 * u][j + 3 * v] == "S"
    )


def part2(data: str) -> int:
    rows = data.splitlines()
    return sum(
        1
        for i in range(len(rows))
        if 0 <= i - 1 and i + 1 < len(rows)
        for j in range(len(rows[0]))
        if 0 <= j - 1 and j + 1 < len(rows[0])
        if rows[i][j] == "A"
        if (rows[i - 1][j - 1] == "M" and rows[i + 1][j + 1] == "S")
        or (rows[i - 1][j - 1] == "S" and rows[i + 1][j + 1] == "M")
        if (rows[i - 1][j + 1] == "M" and rows[i + 1][j - 1] == "S")
        or (rows[i - 1][j + 1] == "S" and rows[i + 1][j - 1] == "M")
    )
